File: backend/src/routes/deliverables.ts
import { Router, Request, Response } from "express";
import { prisma } from "../config/db";
import { z } from "zod";

const router = Router({ mergeParams: true });

const submitDeliverableSchema = z.object({
  submitted_by: z.string().min(1),
  label: z.string().optional().default(""),
  kind: z.string().optional().default("link"),
  url: z.string().optional().default(""),
  file_hash: z.string().optional().nullable(),
});

// GET /api/v1/jobs/:id/deliverables
router.get("/", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id: jobId } = req.params;

    const deliverables = await prisma.deliverables.findMany({
      where: { job_id: jobId },
      orderBy: [
        { milestone_index: "asc" },
        { created_at: "desc" },
      ],
    });

    res.json(deliverables);
  } catch (error) {
    console.error("GET /jobs/:id/deliverables error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

// POST /api/v1/jobs/:id/deliverables
router.post("/", async (req: Request<{ id: string }>, res: Response) => {
  try {
    const { id: jobId } = req.params;
    const data = submitDeliverableSchema.parse(req.body);

    const job = await prisma.jobs.findUnique({
      where: { id: jobId },
      select: { status: true, freelancer_address: true },
    });

    if (!job) {
      return res.status(404).json({ error: `job ${jobId} not found` });
    }

    if (!["funded", "in_progress"].includes(job.status)) {
      return res.status(400).json({ error: `deliverables can only be submitted for funded jobs, not '${job.status}'` });
    }

    if (job.freelancer_address !== data.submitted_by) {
      return res.status(400).json({ error: "only the assigned freelancer can submit deliverables" });
    }

    const nextMilestone = await prisma.milestones.findFirst({
      where: { job_id: jobId, status: "pending" },
      orderBy: { index: "asc" },
    });

    if (!nextMilestone) {
      return res.status(400).json({ error: "all milestones have already been delivered" });
    }

    const result = await prisma.$transaction(async (tx) => {
      const deliverable = await tx.deliverables.create({
        data: {
          job_id: jobId,
          milestone_index: nextMilestone.index,
          submitted_by: data.submitted_by,
          label: data.label,
          kind: data.kind,
          url: data.url,
          file_hash: data.file_hash,
        },
      });

      await tx.jobs.update({
        where: { id: jobId },
        data: { status: "deliverable_submitted" },
      });

      return deliverable;
    });

    res.json(result);
  } catch (error) {
    if (error instanceof z.ZodError) {
      return res.status(400).json({ error: error.issues[0]?.message || "Validation failed" });
    }
    console.error("POST /jobs/:id/deliverables error:", error);
    res.status(500).json({ error: "Internal server error" });
  }
});

export default router;

File: backend/src/routes/deliverables.rs
use axum::{
    extract::{Path, Query, State},
    Json,
};
use uuid::Uuid;

use crate::{
    db::AppState,
    error::{AppError, Result},
    models::{Deliverable, SubmitDeliverableRequest},
    routes::pagination::PaginationQuery,
};

#[tracing::instrument(skip(state, pagination))]
pub async fn list_deliverables(
    State(state): State<AppState>,
    Path(job_id): Path<Uuid>,
    Query(pagination): Query<PaginationQuery>,
) -> Result<Json<Vec<Deliverable>>> {
    let bounds = pagination.bounds();
    let deliverables = sqlx::query_as::<_, Deliverable>(
        r#"SELECT id, job_id, milestone_index, submitted_by, label, kind, url, file_hash, created_at
           FROM deliverables
           WHERE job_id = $1
           ORDER BY milestone_index ASC, created_at DESC, id DESC
           LIMIT $2 OFFSET $3"#,
    )
    .bind(job_id)
    .bind(bounds.limit)
    .bind(bounds.offset)
    .fetch_all(&state.pool)
    .await?;

    Ok(Json(deliverables))
}

#[tracing::instrument(skip(state, req))]
pub async fn submit_deliverable(
    State(state): State<AppState>,
    Path(job_id): Path<Uuid>,
    Json(req): Json<SubmitDeliverableRequest>,
) -> Result<Json<Deliverable>> {
    let (status, freelancer_address): (String, Option<String>) = sqlx::query_as(
        r#"SELECT status, freelancer_address
           FROM jobs
           WHERE id = $1"#,
    )
    .bind(job_id)
    .fetch_optional(&state.pool)
    .await?
    .ok_or_else(|| AppError::NotFound(format!("job {job_id} not found")))?;

    if !matches!(status.as_str(), "funded" | "in_progress") {
        return Err(AppError::BadRequest(format!(
            "deliverables can only be submitted for funded jobs, not '{status}'"
        )));
    }

    if freelancer_address.as_deref() != Some(req.submitted_by.as_str()) {
        return Err(AppError::BadRequest(
            "only the assigned freelancer can submit deliverables".into(),
        ));
    }

    let milestone_index: i32 = sqlx::query_scalar(
        r#"SELECT index
           FROM milestones
           WHERE job_id = $1 AND status = 'pending'
           ORDER BY index ASC
           LIMIT 1"#,
    )
    .bind(job_id)
    .fetch_optional(&state.pool)
    .await?
    .ok_or_else(|| AppError::BadRequest("all milestones have already been delivered".into()))?;

    let deliverable = sqlx::query_as::<_, Deliverable>(
        r#"INSERT INTO deliverables (job_id, milestone_index, submitted_by, label, kind, url, file_hash)
           VALUES ($1, $2, $3, $4, $5, $6, $7)
           RETURNING id, job_id, milestone_index, submitted_by, label, kind, url, file_hash, created_at"#,
    )
    .bind(job_id)
    .bind(milestone_index)
    .bind(req.submitted_by)
    .bind(req.label)
    .bind(req.kind)
    .bind(req.url)
    .bind(req.file_hash)
    .fetch_one(&state.pool)
    .await?;

    sqlx::query("UPDATE jobs SET status = 'deliverable_submitted' WHERE id = $1")
        .bind(job_id)
        .execute(&state.pool)
        .await?;

    Ok(Json(deliverable))
}
